src/zigCompilerProvider.ts

```diff
@@ -1,101 +1,230 @@
+  private buildDiagnostics: vscode.DiagnosticCollection;
+  private astDiagnostics: vscode.DiagnosticCollection;
+  private dirtyChange = new WeakMap<vscode.Uri, boolean>();
+
+  public activate(subscriptions: vscode.Disposable[]) {
+    subscriptions.push(this);
+    this.buildDiagnostics = vscode.languages.createDiagnosticCollection("zig");
+    this.astDiagnostics = vscode.languages.createDiagnosticCollection("zig");
+
+    // vscode.workspace.onDidOpenTextDocument(this.doCompile, this, subscriptions);
+    // vscode.workspace.onDidCloseTextDocument(
+    //   (textDocument) => {
+    //     this.diagnosticCollection.delete(textDocument.uri);
+    //   },
+    //   null,
+    //   subscriptions
+    // );
+
+    // vscode.workspace.onDidSaveTextDocument(this.doCompile, this);
+    vscode.workspace.onDidChangeTextDocument(
+      this.maybeDoASTGenErrorCheck,
+      this
+    );
+  }
+
+  maybeDoASTGenErrorCheck(change: vscode.TextDocumentChangeEvent) {
+    if (change.document.languageId !== "zig") return;
+    if (change.document.isClosed) {
+      this.astDiagnostics.delete(change.document.uri);
+    }
 
-    public activate(subscriptions: vscode.Disposable[]) {
-        subscriptions.push(this);
-        this.diagnosticCollection = vscode.languages.createDiagnosticCollection("zig");
+    this.doASTGenErrorCheck(change);
 
-        vscode.workspace.onDidOpenTextDocument(this.doCompile, this, subscriptions);
-        vscode.workspace.onDidCloseTextDocument((textDocument) => {
-            this.diagnosticCollection.delete(textDocument.uri);
-        }, null, subscriptions);
+    if (!change.document.isUntitled) {
+      let config = vscode.workspace.getConfiguration("zig");
+      if (
+        config.get<boolean>("buildOnSave") &&
+        this.dirtyChange.has(change.document.uri) &&
+        this.dirtyChange.get(change.document.uri) !== change.document.isDirty &&
+        !change.document.isDirty
+      ) {
+        this.doCompile(change.document);
+      }
 
-        vscode.workspace.onDidSaveTextDocument(this.doCompile, this);
+      this.dirtyChange.set(change.document.uri, change.document.isDirty);
     }
-
-    public dispose(): void {
-        this.diagnosticCollection.clear();
-        this.diagnosticCollection.dispose();
+  }
+
+  public dispose(): void {
+    this.buildDiagnostics.clear();
+    this.astDiagnostics.clear();
+    this.buildDiagnostics.dispose();
+    this.astDiagnostics.dispose();
+  }
+
+  private _doASTGenErrorCheck(change: vscode.TextDocumentChangeEvent) {
+    let config = vscode.workspace.getConfiguration("zig");
+    const textDocument = change.document;
+    if (textDocument.languageId !== "zig") {
+      return;
+    }
+    const zig_path = config.get("zigPath") || "zig";
+    const cwd = vscode.workspace.getWorkspaceFolder(textDocument.uri).uri
+      .fsPath;
+
+    let childProcess = cp.spawn(
+      zig_path as string,
+      ["astgen", "--errors-only", "--stdin"],
+      { cwd }
+    );
```

Comment:
```suggestion
    let childProcess = cp.spawn(zig_path as string, ["ast-check"], { cwd });
```
Andrew made some changes to your patch.